1. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

2. Set Path

In [2]:
# Setting 'path' as Instacart Achievement Folder 
path = r'C:\Users\Josh Wattay\anaconda3\Instacart Basket Analysis'

3. Import Data Sets

In [3]:
#creating dataframe for products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False) 

In [4]:
#creating dataframe for wrangled orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

4. Create a Dataframe to Practice Fixing Mixed-Type Data Columns

In [6]:
#Create a dataframe
df_test = pd.DataFrame()

In [7]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
#Test Output for new dataframe
df_test.head()

mix
0     a
1     b
2     1
3  True

5. Code to Check for Mixed-Type Data Columns

In [10]:
#check for mixed-type data
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test
             [[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


6. Changing Data Type

In [13]:
#Changing Data Type in mix column to String
df_test['mix']=df_test['mix'].astype('str')

7. Missing Values

In [19]:
#Using isnull to find missing values in products data
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
#Creating new dataframe to observe null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

8. Addressing Missing Values

In [ ]:
#Code to Impute the Mean (do not run)
df.describe() to find the mean THEN
df['column with missings'].fillna(mean value, inplace = True)

In [ ]:
#Code to Impute the Median (do not run)
df_prods.median() to find the median THEN
df['column with missings'].fillna(median value, inplace = True)

9. Determining Shape of Data before Proceeding

In [22]:
#Shape of data will help us determine what proportion of the rows are missing product names
df_prods.shape

(49693, 5)

In [23]:
#only 16 of 49693 rows are missing product names

10. Creating Clean Products Dataset

In [26]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [27]:
#running shape to verify that 16 rows have been removed in the clean data subset
df_prods_clean.shape

(49677, 5)

11. Alternative Method for Dropping Missing Values

In [ ]:
df_prods.dropna(inplace = True)
#do not run

12. Method for Dropping NaN from a Specific Column

In [ ]:
df_prods.dropna(subset = ['product_name'], inplace = True)
#do not run

13. Duplicates

In [28]:
#command to look for duplicates in dataframe
df_dups = df_prods_clean [df_prods_clean.duplicated()]

In [29]:
#creates a dataframe subsetting duplicates from the df_prods_clean dataframe

In [30]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [31]:
df_prods_clean.shape
#verify shape prior to dropping duplicates

(49677, 5)

In [32]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
#dropping duplicates and creating new clean subset

In [33]:
df_prods_clean_no_dups.shape
#verify shape after dropping duplicates

(49672, 5)

14. Exporting Clean Data Set

In [34]:
#Exporting df_prods_clean_no_dups into prepared data folder.
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

15. EXERCISE 4.5 STEP 1

In [35]:
#This step was completed at the top of this notebook, please reference 1. Import Libraries and 3. Import Data Sets

16. EXERCISE 4.5 STEP 2

In [37]:
#Changing format in Pandas to 2 decimal places for ease of reference
pd.set_option('display.float_format', '{:.2f}'.format)

In [39]:
#Running df.describe() on the Wrangled Orders Data
summary_stats = df_ords.describe()
summary_stats

Unnamed: 0   order_id    user_id  order_number  orders_day_of_week  \
count  3421083.00 3421083.00 3421083.00    3421083.00          3421083.00   
mean   1710541.00 1710542.00  102978.21         17.15                2.78   
std     987581.74  987581.74   59533.72         17.73                2.05   
min          0.00       1.00       1.00          1.00                0.00   
25%     855270.50  855271.50   51394.00          5.00                1.00   
50%    1710541.00 1710542.00  102689.00         11.00                3.00   
75%    2565811.50 2565812.50  154385.00         23.00                5.00   
max    3421082.00 3421083.00  206209.00        100.00                6.00   

       order_hour  days_since_prior_order  
count  3421083.00              3214874.00  
mean        13.45                   11.11  
std          4.23                    9.21  
min          0.00                    0.00  
25%         10.00                    4.00  
50%         13.00                    7.00  
75%         16.00                   15.00  
max         23.00                   30.00

Observations of Wrangled Orders Descriptive Statistics.

Order Hour - The minimum order hour is 0 and the maximum is 23. This is somewhat counterintuitive for those using non-military or non-24-hour methods of tracking time. However, for those that do, 0000 is midnight and 2300 is 11pm, therefore this is possibly ok, but there may be a better way to display the hours if we go from 1 through 24. 

Order Number - There were 3,421,083 unique order id's and therefore the same amount of orders. However, the minimum order number is 1 and the maximum is 100. I interpret this to mean that the order number is specific to the user id. User ID 1 may have had 4 orders and user ID 2 may have had 100 orders. It appears the average number of orders per user was 17.15.

Orders Day of Week - Similar to Order Hour, this variable has somewhat counterintuitive minimum (0) and maximum (6) values. It is hard to determine whether 0 is Sunday or Monday, and likewise whether 6 is Friday, Saturday, or Sunday. I believe changing this from a range of 1-7 would be more intuitive with 1 being Monday and 7 being Sunday.

17. EXERCISE 4.5 STEP 3

Note: I used dataframe.map instead of dataframe.applymap because I received a FutureWarning that .applymap is deprecating and received a recommendation to use dataframe.map instead.

In [42]:
#checking for mixed data-type in df_ords using a for loop command
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords
             [[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

Since no columns were printed, this for loop is indicating there are no mixed data-type columns in the df_ords dataframe.

18. EXERCISE 4.5 STEP 4

No mixed data type columns were located, however, if they were located, I would use the df_ords['column_name']=df_ords['column_name'].astype('str') to change the data type to string OR df_ords['column_name']=df_ords['column_name'].astype('int64') to change the data type to integer.

19. EXERCISE 4.5 STEP 5

In [44]:
#Checking for Missing Values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour                     0
days_since_prior_order    206209
dtype: int64

There are 206,209 missing values in the days_since_prior_order column. 
I believe these may have been a User's first order and would like to check with the stakeholders to ensure that this is the case prior to proceeding further with the analysis. 
Days since prior order would be non-applicable to someone who is ordering for the first time.

20. EXERCISE 4.5 STEP 6

In [45]:
#Determining the shape of df_ords to compare amount of missing values in days since prior order to total amount of records
df_ords.shape

(3421083, 8)

Creating subsets of df_ords_first_orders (null values) and df_ords_returning (no null values) to subset the first orders and analyze separately from remaining 3,214,874 records for repeat customers

In [47]:
df_ords_returning = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [48]:
df_ords_first_orders = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [51]:
df_ords_returning.shape
#verifying based on df.shape

(3214874, 8)

In [52]:
df_ords_first_orders.shape
#verifying based on df.shape

(206209, 8)

I used this method because I don't want to ever overwrite data in order to allow my peers, stakeholders, and superiors to be able to audit my work at all times. I would rather create two subsets while awaiting further clarification from the stakeholder as to why the days since prior order value is missing from so many records. 

21. EXERCISE 4.5 STEP 7

In [53]:
#Checking for duplicates in the df_ords data
df_dups_ords = df_ords [df_ords.duplicated()]

In [54]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour, days_since_prior_order]
Index: []

There were no duplicate records detected. This is most likely due to the fact that each order has two unique identifiers, an order id and a user id. More unique identifiers in a record likely decreases the probability of duplicate records, depending on the method for data collection.

22. EXERCISE 4.5 STEP 8

There were no duplicate records detected. If there were, I would create a subset of the duplicate data and another subset of clean orders data to keep a clean subset and a duplicate subset without overwriting the original dataset.

23. EXERCISE 4.5 STEP 9

In [55]:
#Exporting Clean Orders Data as df_orders_checked csv file. 
df_ords_returning.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

Please reference Markdown Section 14 for the upload of clean products data.